packages/cursorless-engine/src/languages/TreeSitterQuery/treeSitterQueryCache.ts

```diff
@@ -0,0 +1,72 @@
+import type { Position, TextDocument } from "@cursorless/common";
+import type { QueryMatch } from "./QueryCapture";
+
+export class Cache {
+  private documentUri: string = "";
+  private documentVersion: number = -1;
+  private documentLanguageId: string = "";
+  private startPosition: Position | undefined;
+  private endPosition: Position | undefined;
+  private matches: QueryMatch[] = [];
+
+  isValid(
+    document: TextDocument,
+    startPosition: Position | undefined,
+    endPosition: Position | undefined,
+  ) {
+    return (
+      this.documentUri === document.uri.toString() &&
+      this.documentVersion === document.version &&
+      this.documentLanguageId === document.languageId &&
+      this.startPosition === startPosition &&
+      this.endPosition === endPosition
+    );
+  }
+
+  update(
+    document: TextDocument,
+    startPosition: Position | undefined,
+    endPosition: Position | undefined,
+    matches: QueryMatch[],
+  ) {
+    this.documentUri = document.uri.toString();
+    this.documentVersion = document.version;
+    this.documentLanguageId = document.languageId;
+    this.startPosition = startPosition;
+    this.endPosition = endPosition;
+    this.matches = matches;
+  }
+
+  get(): QueryMatch[] {
```

Comment:
Note to self: I was surprised we are caching a lower level tree sitter object rather than something that we have post processed but it looks like we use this later on in other parts of the codebase, so we need to keep it around. It's probably not worth the work to change this.

https://github.com/cursorless-dev/cursorless/blob/1bffae54938cc5702000d72812bc04ea1ac0dab8/packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/TreeSitterScopeHandler/TreeSitterIterationScopeHandler.ts#L40

https://github.com/cursorless-dev/cursorless/blob/main/packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/TreeSitterScopeHandler/captureUtils.ts#L16-L36